Episodes: Profiling Intro/Functions/Lines/Conclusion (#5)
RSE-Sheffield/pando-python: episodes/profiling-introduction.md

:::::::::::::::::::::::::::::::::::::: questions

- Why should you profile your code?
- How should you choose which type of profiler to use?
- Which test case should be profiled?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- explain the benefits of profiling code and different types of profiler
- identify the appropriate Python profiler for a given scenario
- explain how to select an appropriate test case for profiling and why

::::::::::::::::::::::::::::::::::::::::::::::::

## Introduction

<!-- Profiling is (what) -->
Performance profiling is the process of analysing and measuring the performance of a program or script, to understand where time is being spent during execution.

<!-- It can be used for (where) -->
Profiling is useful when you have written any code that will be running for a substantial period of time.
As your code grows in complexity, it becomes increasingly difficult to estimate where time is being spent during execution.

> Profiling allows you to narrow down where the time is being spent, to identify whether this is of concern or not.

Author: ```suggestion
Profiling allows you to narrow down where the time is being spent, to prioritise optimisation efforts.
```
